Next library

marksmall opened the pull request:
I've updated the library to contain the `Header` and `Footer` components from the DRL app, so these components can be shared among many apps.

This was no simple task, it involved working out how to preserve lines such as:

* 'use client';
* 'use server';

These were being stripped out by the vite/rollup build.

I also added the ability to export images from the app, that we may want to share as well as some common types e.g. `Credentials`.

I also updated the ErrorText component to have the role `alert` so it is more semantically correct for accessibility.

marksmall committed:
fix(docs): Fix typo where wrong filename specified

The docs stated to look at the `src/main.ts` file but that doesn't
exist, really it should have been `src/index.ts`. I have now fixed it.

marksmall committed:
fix(project): Add necessary dependencies

I've added dependencies used by this library so that we can have
next-based components in the library e.g. custom `next/link` component.

I also added some vite plugins to manage static image we may want
to provide as well as SVG management. It's not as simple as adding
images to the repo, unless they are only used internally, if we want
to be able to import them into app code, then they need to these
plugins.

marksmall committed:
feat(assets): Add ability to export assets for apps to use

I've added the ability to export assets such as fonts and images so
that an app can import them to be used as if they were local to the
app.

marksmall committed:
fix(project): Ensure we include src directory for typescript

I updated the include section while testing fixes for other issues
being seen. Not entirely sure why this is necessary, but the library
works and I don't want to remove it.

marksmall committed:
fix(project): Ensure `use client` and `use server` are honoured

Once of the big changes I've made is this, before vite/rollup was
stripping out this code, so that when I used the component in an
app, it complained that it needed `'use client';`, for example. I
couldn't understand at first and only with the help of ChaGPT did
I figure out it was vite/rollup. So now these lines should be preserved
in any module that uses them.

marksmall committed:
fix(project): Add types used by header/footer

I've added these types as they are used in the Header/Footer components
that will be committed soon.

marksmall committed:
feat(components): Add dropdown/select menu component

This component is to be used by the forth-coming Header component,
I want to ensure though that I don't try to commit too much in one.

marksmall committed:
fix(tests): Add custom @testing-library renderer

I actually use this in the code I previously committed, I should have
committed this first but got it wrong, so here it is. It is a custom
renderer, so that we can wrap components that need `providers` etc
to test.

marksmall committed:
feat(components): Add link components used in the Header/Footer

I've added a couple of components that are currently used in the
Header/Footer, but will also be required within our apps also, so
I export them so they can be used.

marksmall committed:
feat(components): Add SVG images as normal react components

There is little benefit to exposing these SVGs as images, when used
in `next/image` there is no optimisation involved as that only really
relates to PNGs/GIFs/JPGs etc. Rather than complicate code by importing
the SVG, then wrapping it in an image, I felt it was easier just to
create components from them and treat them just like any other JSX.

NOTE: I have not re-exported these from the library as right now it
      is not necessary, but they could be if that changes.

marksmall committed:
feat(layout): Add Header/Footer components

I have now added the complex Header/Footer components, which re-use
some of the components I have previously committed and fixes to the
vite/rollup configs.

marksmall committed:
fix(components): Add role to ErrorText component

I've added a role to this element to make it more accessible.

marksmall committed:
fix(types): Export types so they can be re-used in apps

I've added some common types, so they can be re-used in our apps.

marksmall committed:
fix(project): Add globals that might be used

I've added `globals.browser` as we are already using the `document`
and `window` global objects in the code.

I also added some node-specific overrides for any server only files.

marksmall committed:
feat(project): Run script after dependency installation

There are some manual commands needed for the repo, to manage a local
GIT message on commits and setting up husky hooks etc. Instead of
relying on devs remembering, I've wrapped these in a shell script and
call it from the `postinstall` life-cycle script in package.json.
This ensures the commands are run after dependencies installed, which
is generally one of the first things a dev does.

DexterHill0 commented on src/components/layout/footer/Footer.tsx:
```diff
@@ -0,0 +1,26 @@
+        grid-rows-[min-content_min-content_min-content] md:grid-rows-[min-content_min-content]"
+    >
+      <nav className="row-start-1 col-start-1 md:col-span-2 w-full pb-2">
+        <h2 className="sr-only">Support Links</h2>
```
something for later, but as we removed headers from the Header we might want to remove this one and change it to a different semantic element.

marksmall commented on src/components/dropdown/useDropdownMenu.ts:
```diff
@@ -0,0 +1,228 @@
+    },
+
+    itemProps: Array.from({ length: itemCount }, (_, itemIndex) => ({
+      key: itemIndex,
+
```
Not sure about this, your wrote this component @DexterHill0, what is your opinion? I agree using `index` as a key is bad, just not sure if we can delay this to usage of the items in the app or not.

DexterHill0 commented on src/components/layout/header/Header.tsx:
```diff
@@ -0,0 +1,72 @@
+  appName: string;
+};
+
+export const HEADER_DEFAULT_NAV_LINKS = [{ label: 'API', url: '/api-docs' }];
```
just a reminder to make the nav links a prop

marksmall commented on src/components/layout/header/HeaderNavClient.tsx:
```diff
@@ -0,0 +1,68 @@
+    } else {
+      setLinks(navLinks);
+    }
+  }, [revision]);
```
This is something we should look at separately, but is the library using our latest config and have the `react-hooks` eslint plugin installed.

DexterHill0 commented on src/components/layout/header/HeaderNavClient.tsx:
```diff
@@ -0,0 +1,68 @@
+
+  useEffect(() => {
+    if (revision) {
+      setLinks([{ label: 'Assets Library', url: `/revision/${revision}/categories` }, ...navLinks]);
```
This is what I was talking about where in DRL it changes the header URLs depending on what page you are on. This is hopefully doable using route groups instead.

Not sure what to do with this code, as it's only applicable to DRL but if we remove it DRL will break

DexterHill0 replied:
I think that might be for the best as we don't want to include this code in non-DRL apps

DexterHill0 commented on src/components/links/ExternalLink.tsx:
```diff
@@ -0,0 +1,28 @@
+  url: string;
+  className?: string;
+  children: ReactNode;
+  [key: string]: unknown;
```
possible future change, but now that it's in the UI kit, we can use the `ExtendProps` utility instead of this. Something like:
```ts
type Props = {
  url: string;
  className?: string;
  children: ReactNode;
}

type ExternalLinkProps = ExtendProps<'a', Props>;
```
will also improve the intellisense

marksmall replied:
Done.

DexterHill0 commented on src/components/links/Link.tsx:
```diff
@@ -0,0 +1,27 @@
+  url: string | object;
+  className?: string;
+  children: ReactNode;
+  [key: string]: unknown;
```
Same comment as the `Link` component, although because this is a next link it will probably have to be `ExtendProps<typeof NextLink, ...>`

marksmall replied:
Done.

DexterHill0 reviewed (changes requested):
Few comments, but they could all easily be left for later. Otherwise everything looks sensible

marksmall committed:
fix(project): Update to latest eslint config

We released a new version of our ESLint config, so have now updated
the library to use it.

marksmall committed:
fix(project): Update to latest Node LTS version

I've updated the library to use the latest Node LTS version.

marksmall committed:
fix(components): Update components after ESLint config update

I've updated all components that needed updating after upgrading the
version of our shared ESLint configs.

dspgithub-ops commented:
:tada: This PR is included in version 1.4.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@tpzdsp/tpz-dsp3-ui-kit/v/1.4.0)
- [GitHub release](https://github.com/DSP3-apps/tpz-dsp3-ui-kit/releases/tag/v1.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
